pkg/storegateway/series_refs_streaming.go

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package storegateway
+
+type iteratorFactory func(strategy seriesIteratorStrategy) iterator[seriesChunkRefsSet]
+
+// chunksStreamingCachingSeriesChunkRefsSetIterator is an iterator used while streaming chunks from store-gateways to queriers.
+//
+// It wraps another iterator that does the actual work. If that iterator is expected to produce only a single batch,
+// this iterator caches that batch for the chunks streaming phase, to avoid repeating work done during the series label sending phase.
+type chunksStreamingCachingSeriesChunkRefsSetIterator struct {
+	strategy             seriesIteratorStrategy
+	postingsSetsIterator *postingsSetsIterator
+	factory              iteratorFactory
+	it                   iterator[seriesChunkRefsSet]
+
+	expectSingleBatch                    bool
+	inChunksStreamingPhaseForSingleBatch bool
+	haveCachedBatch                      bool // It's possible that we expected a single batch to be returned, but the batch was filtered out by the inner iterator.
+
+	currentBatchIndex int // -1 after beginning chunks streaming phase, 0 when on first and only batch, 1 after first and only batch
```

Comment:
this seems to be only incremented in the chunks streaming phase and only used to indicate that the single batch has been returned. Can we replace it with a boolean? `returnedCachedBatch: true|false` or maybe `consumedCachedBatch`?

Reply:
I'm not sure we can replace this with a boolean. `Next` will be called twice in the case where we're returning a single cached batch for the chunks streaming phase. 

On the first call, `Next` advances to the first batch, and on the second call, `Next` advances to the end of the iterator. 

We also can't track whether we've returned the cached set in `At`, as it's valid to call `At` multiple times between `Next` calls.

We might be able to replace this with two booleans (`haveAdvancedToCachedBatch`, `haveAdvancedPastCachedBatch`), but I'm not sure that's any better. Open to other opinions or ideas though.

Reply:
you're right. Happy to keep the counter